dds/DCPS/NetworkConfigPublisher.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ *
+ *
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include "DCPS/DdsDcps_pch.h" //Only the _pch include should start with DCPS/
+
+#include "NetworkConfigPublisher.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+NetworkInterfaces NetworkConfigPublisher::add_listener(NetworkConfigListener_rch listener)
+{
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, NetworkInterfaces());
+  listeners_.insert(listener);
+  return network_interfaces_;
+}
+
+void NetworkConfigPublisher::remove_listener(NetworkConfigListener_rch listener)
+{
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  listeners_.erase(listener);
+}
+
+NetworkInterfaces NetworkConfigPublisher::get() const
+{
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, NetworkInterfaces());
+  return network_interfaces_;
+}
+
+void NetworkConfigPublisher::add_interface(const NetworkInterface& nic)
+{
+  Listeners listeners;
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+    listeners = listeners_;
+    network_interfaces_.push_back(nic);
+  }
+  for (Listeners::const_iterator pos = listeners.begin(), limit = listeners.end(); pos != limit; ++pos) {
+    (*pos)->add_interface(nic);
+  }
+}
+
+void NetworkConfigPublisher::remove_interface(int index)
+{
+  NetworkInterface nic;
+  Listeners listeners;
+  bool publish = false;
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+    listeners = listeners_;
+    NetworkInterfaces::iterator pos = std::find(network_interfaces_.begin(), network_interfaces_.end(), index);
+    if (pos != network_interfaces_.end()) {
+      nic = *pos;
+      network_interfaces_.erase(pos);
+      publish = true;
+    }
+  }
+  if (publish) {
+    while (!nic.addresses.empty()) {
```

Comment:
Is this part of the NCP -> Listener interaction needed?  Wouldn't a Listener know from `remote_interface` that all addresses belonging to that interface are effectively removed?

Reply:
Yes.  But I found that using it is slightly easier if we don't use composite events.  For most cases, listeners won't even care about the interfaces and only care about changes in addresses.